examples/golang/components/http-password-checker/main.go

```diff
@@ -0,0 +1,79 @@
+//go:generate go run github.com/bytecodealliance/wasm-tools-go/cmd/wit-bindgen-go generate --world hello --out gen ./wit
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	gopasswordvalidator "github.com/wagslane/go-password-validator"
+	"go.wasmcloud.dev/component/net/wasihttp"
+)
+
+type CheckRequest struct {
+	Value string `json:"value"`
+}
+
+type CheckResponse struct {
+	Valid  bool `json:"valid"`
+	Length int  `json:"length"`
+}
+
+func init() {
+	// Register the handleRequest function as the handler for all incoming requests.
+	wasihttp.HandleFunc(handleRequest)
+}
+
+func checkHandler(w http.ResponseWriter, r *http.Request) {
+	var req CheckRequest
+	defer r.Body.Close()
+
+	b, err := io.ReadAll(r.Body)
+	if err != nil {
+		errResponseJSON(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	if err := json.Unmarshal(b, &req); err != nil {
+		errResponseJSON(w, http.StatusBadRequest, fmt.Sprintf("error with json input: %s", err.Error()))
+		return
+	}
+
+	err = gopasswordvalidator.Validate(req.Value, 60)
+	if err != nil {
+		errResponseJSON(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	resp := CheckResponse{Valid: true, Length: len(req.Value)}
+	respJSON, err := json.Marshal(resp)
+	if err != nil {
+		errResponseJSON(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(respJSON)
+}
+
+func handleRequest(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		errResponseJSON(w, http.StatusMethodNotAllowed, "Method not allowed")
+		return
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/check", checkHandler)
+	mux.ServeHTTP(w, r)
+}
```

Comment:
Technically you could get rid of `handleRequest` altogether by moving the check for MethodPost
```go
func checkHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		errResponseJSON(w, http.StatusMethodNotAllowed, "Method not allowed")
		return
	}
	var req CheckRequest
	defer r.Body.Close()
```
at the top of the `checkHandler` function, and moving the `mux` setup into `init()`:
```go
func init() {
	// Register the ServeMux as the handler for all incoming requests.
	mux := http.NewServeMux()
	mux.HandleFunc("/check", checkHandler)
	wasihttp.Handle(mux)
}
```

Reply:
You know what's really weird -- I didn't find that *any* of the muxes (`http.ServeMux`, `gorilla/mux`, etc) worked properly with `init` or the handler in this way (I tried this a few days ago working on stuff for WASMCON 2024).

I couldn't pin down *why* it wasn't working -- but the paths just wouldn't parse (the 404 handler would be the only thing returned) -- I stopped before descending into match machinery hell but I do want to note it's not as easy as it seems. Maybe you'll have more luck here.

Reply:
It was fixed in TinyGo 0.34.0.

Reply:
Yeah, I've run this code and it very much works

Reply:
Oh that's great, I'm running an old version of TinyGo it seems -- I'm on 0.33 which explains it!

Do we want to change this code to be that? Also while we're here it might be worth adding a test other than the default wadge one -- IIRC it just checks `/`